compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,424 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.dnd
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.graphics.toPainter
+import androidx.compose.ui.layout.LayoutCoordinates
+import androidx.compose.ui.layout.boundsInWindow
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.window.LocalWindow
+import androidx.compose.ui.window.density
+import java.awt.Image
+import java.awt.Point
+import java.awt.Window
+import java.awt.datatransfer.DataFlavor
+import java.awt.datatransfer.DataFlavor.selectBestTextFlavor
+import java.awt.datatransfer.Transferable
+import java.awt.dnd.DnDConstants
+import java.awt.dnd.DropTarget
+import java.awt.dnd.DropTargetDragEvent
+import java.awt.dnd.DropTargetDropEvent
+import java.awt.dnd.DropTargetEvent
+import java.awt.dnd.DropTargetListener
+import java.awt.image.BufferedImage
+import java.io.File
+
+/**
+ * Represent data types drag and dropped to an application from outside.
+ */
+@ExperimentalComposeUiApi
+sealed interface DropData {
+    /**
+     * Represents list of files drag and dropped to an application in a raw [java.net.URI] format.
+     */
+    data class FilesList(val rawUris: List<String>) : DropData
+
+    /**
+     * Represents an image drag and dropped to an application.
+     */
+    data class Image(val painter: Painter) : DropData
+
+    /**
+     * Represent text drag and dropped to an application.
+     *
+     * @param mimeType mimeType of the [content] such as "text/plain", "text/html", etc.
+     */
+    data class Text(val content: String, val mimeType: String?) : DropData
+}
+
+/**
+ * Adds detector of external drag and drop (e.g. files DnD from Finder to an application)
+ *
+ * @param onDragStart will be called when the pointer with external content entered the component.
+ * @param onDrag will be called for all drag events inside the component.
+ * @param onDrop is called when the pointer is released with [DropData] the pointer held.
+ * @param onDragCancel is called if the pointer exited the component bounds or unknown data was dropped.
```

Comment:
Hm, weird that IDEA doesn't move it to the new line. Where do you see this into styleguide?

Reply:
Not sure; @igordmn told me.

Reply:
This was worked perfectly in Android Studio in AOSP, but we have bugs in our fork, and this limit doesn't always apply.

There is an idea to increase it for our fork. It will complicate upstreaming into AOSP, but that is another dicsussion.

Reply:
Check your settings in IDEA, if we you will find that you don't use `Project`, switch to it, and then comply to it:
![image](https://user-images.githubusercontent.com/5963351/224427268-61a0223b-48bd-4897-b5cc-224f63ca4cb8.png)

Reply:
I use the project one, weird